File: services/lake-query/src/lake_query/connection.py
"""DuckDB connection helper for querying Parquet on S3."""

import os
import duckdb
from typing import Optional


def get_duckdb_conn(
    lake_bucket: Optional[str] = None,
    read_only: bool = False,
) -> duckdb.DuckDBPyConnection:
    """
    Get a configured DuckDB connection for querying the data lake.

    Configures:
    - httpfs extension for S3 access
    - S3 endpoint, credentials, and URL style from environment variables
    - Hive partitioning for efficient partition pruning

    Args:
        lake_bucket: S3 bucket name (defaults to LAKE_BUCKET env var)
        read_only: whether to open connection in read-only mode

    Returns:
        Configured DuckDB connection
    """
    conn = duckdb.connect(":memory:", read_only=read_only)

    # Install and load httpfs extension
    conn.execute("INSTALL httpfs")
    conn.execute("LOAD httpfs")

    # Configure S3 from environment variables
    endpoint = os.getenv("AWS_ENDPOINT_URL")
    access_key = os.getenv("AWS_ACCESS_KEY_ID", "")
    secret_key = os.getenv("AWS_SECRET_ACCESS_KEY", "")
    region = os.getenv("AWS_REGION", "us-east-1")
    use_ssl = os.getenv("LAKE_S3_USE_SSL", "false").lower() == "true"
    url_style = os.getenv("LAKE_S3_URL_STYLE", "path")

    if endpoint:
        # Remove scheme from endpoint (DuckDB expects just host:port)
        ep_clean = endpoint.replace("https://", "").replace("http://", "")
        conn.execute(f"SET s3_endpoint='{ep_clean}'")

    conn.execute(f"SET s3_use_ssl={'true' if use_ssl else 'false'}")
    conn.execute(f"SET s3_access_key_id='{access_key}'")
    conn.execute(f"SET s3_secret_access_key='{secret_key}'")
    conn.execute(f"SET s3_url_style='{url_style}'")
    conn.execute(f"SET s3_region='{region}'")

    # Hive partitioning for efficient partition pruning
    conn.execute("SET hive_partitioning=true")

    return conn


def lake_uri(category: str, dataset: str, bucket: Optional[str] = None) -> str:
    """
    Build an S3 URI for a dataset in the data lake.

    Args:
        category: data category (equities, indices, derivatives, etc.)
        dataset: specific dataset name (bhavcopy, history, etc.)
        bucket: S3 bucket name (defaults to LAKE_BUCKET env var)

    Returns:
        S3 URI prefix for the dataset, e.g. s3://destiny-lake/warehouse/equities/bhavcopy
    """
    bucket = bucket or os.getenv("LAKE_BUCKET", "destiny-lake")
    return f"s3://{bucket}/warehouse/{category}/{dataset}"

File: packages/duckdb-lake/src/connection.ts
/**
 * Shared DuckDB connection for querying Parquet on S3 via httpfs.
 *
 * Singleton pattern: one connection per process.
 * Initializes S3 credentials from environment variables on first use.
 */

import Database from "@duckdb/node-api";

let instance: Database | null = null;
let isInitialized = false;

/**
 * Get or create the singleton DuckDB connection.
 *
 * On first call, configures:
 * - INSTALL/LOAD httpfs extension
 * - S3 endpoint, credentials, and URL style from env vars
 * - Memory and threading limits
 */
export async function getConnection(): Promise<Database> {
  if (instance && isInitialized) {
    return instance;
  }

  const db = new Database();

  try {
    // Install and load httpfs extension
    await db.run("INSTALL httpfs");
    await db.run("LOAD httpfs");

    // Configure S3 from environment variables
    const endpoint = process.env.AWS_ENDPOINT_URL;
    const accessKey = process.env.AWS_ACCESS_KEY_ID || "test";
    const secretKey = process.env.AWS_SECRET_ACCESS_KEY || "test";
    const region = process.env.AWS_REGION || "us-east-1";
    const useSsl = process.env.LAKE_S3_USE_SSL === "true";
    const urlStyle = process.env.LAKE_S3_URL_STYLE || "path";

    if (endpoint) {
      // Remove scheme from endpoint (DuckDB expects just host:port)
      const epClean = endpoint.replace(/https?:\/\//, "");
      await db.run(`SET s3_endpoint='${epClean}'`);
    }

    await db.run(`SET s3_use_ssl=${useSsl ? "true" : "false"}`);
    await db.run(`SET s3_access_key_id='${accessKey}'`);
    await db.run(`SET s3_secret_access_key='${secretKey}'`);
    await db.run(`SET s3_url_style='${urlStyle}'`);
    await db.run(`SET s3_region='${region}'`);

    // Hive partitioning
    await db.run("SET hive_partitioning=true");

    // Memory and threading tuning
    const memLimit = process.env.DUCKDB_MEMORY_LIMIT || "2GB";
    const threads = process.env.DUCKDB_THREADS || "4";

    await db.run(`SET memory_limit='${memLimit}'`);
    await db.run(`SET threads=${threads}`);

    instance = db;
    isInitialized = true;

    return db;
  } catch (error) {
    console.error("Failed to initialize DuckDB connection:", error);
    throw error;
  }
}

/**
 * Close the connection (called on process shutdown).
 */
export async function closeConnection(): Promise<void> {
  if (instance) {
    // DuckDB Node API doesn't have explicit close, but we can set to null
    instance = null;
    isInitialized = false;
  }
}

/**
 * Reset the connection (useful for testing).
 */
export function resetConnection(): void {
  instance = null;
  isInitialized = false;
}
